### PR TITLE
ENH: Add action for fetching the CheckV database

### LINKED ISSUE #2: ENH: Add action for fetching the CheckV database

### PR DESCRIPTION
**Pull request by ChristosMatzoros**
This PR closes #2  .

- Adds action to fetch CheckV database.

### Set up the environment
 ```shell
mamba create -n q2-viromics -c conda-forge -c bioconda -c https://packages.qiime2.org/qiime2/2024.10/metagenome/passed/ -c defaults q2cli q2-types checkv pyhmmer
```

###  Activate q2-viromics environment
```shell
conda activate q2-viromics
```

### Clone the repo and checkout the PR branch
```bash
git clone https://github.com/bokulich-lab/q2-viromics.git
cd q2-viromics
git fetch origin pull/6/head:pr-6
git checkout pr-6
pip install -e .
```

### Import database
```bash
# It may take some time!
qiime viromics checkv-fetch-db  --o-database checkv_db.qza --verbose
```

### CODE REVIEW

File: q2_viromics/_utils.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    Args:
+    file_name (str): The name of the file or the relative path to the file.
+
+    Returns:
+    str: The absolute path of the file.
+    """
+
+    return os.path.abspath(file_name)
+
```
Comment (VinzentRisch) [inline q2_viromics/_utils.py]: Why is this here? You are not using this function. It looks like you copied it from another plugin. 
Also whats wrong with just using `os.path.abspath(file_name)` directly?

File: q2_viromics/_utils.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    Returns:
+    bool: True if the directory was created, False if it already existed.
+    """
+
+    try:
+        os.makedirs(path)
+        return True
+    except FileExistsError:
+        return False
+
```
Comment (VinzentRisch) [inline q2_viromics/_utils.py]: Same here why do you have this? You are not using it.
this is also the same as just using `os.makedirs(path, exists_ok: True)`

File: q2_viromics/_utils.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    # Pipe the output of cmd2 to cmd3, but this time manage the output file ourselves
+    with open(outfile_path, "wb") as outfile:
+        p3 = subprocess.Popen(cmd3, stdin=p2.stdout, stdout=outfile)
+        p2.stdout.close()  # Allow p2 to receive a SIGPIPE if p3 exits
+
+    # Wait for the processes to complete
+    p1.wait()
+    p2.wait()
+    p3.wait()
+
```
Comment (VinzentRisch) [inline q2_viromics/_utils.py]: Same with this it is not used anywhere in the plugin.

File: q2_viromics/_utils.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+def _construct_param(arg_name):
+    """Converts argument name into a command line parameter."""
+    return f'--{arg_name.replace("_", "-")}'
+
```
Comment (VinzentRisch) [inline q2_viromics/_utils.py]: Same here

File: q2_viromics/_utils.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    """
+    processed_args = []
+    for arg_key, arg_val in params.items():
+        if not arg_val:
+            continue
+        else:
+            processed_args.extend(processing_func(arg_key, arg_val))
+    return processed_args
+
```
Comment (VinzentRisch) [inline q2_viromics/_utils.py]: Also this one is not used

File: q2_viromics/_utils.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+def _get_sample_from_path(fp):
+    """Extracts sample name from a contig's file path."""
+    return os.path.basename(fp).rsplit("_contigs.fa", maxsplit=1)[0]
```
Comment (VinzentRisch) [inline q2_viromics/_utils.py]: This is also not used.

File: q2_viromics/_version.py
Comment (VinzentRisch) [inline q2_viromics/_version.py:1]: Why do you need this file?
I mean you did not write this yourself, right?
Why not import it somehow?

File: q2_viromics/checkv_fetch_db.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        raise Exception(
+            "An error was encountered while running CheckV download_database, "
+            f"(return code {e.returncode}), please inspect "
+            "stdout and stderr to learn more."
+        )
+
+
+# Fetch the Virsorter2 database
+def checkv_fetch_db() -> CheckVDbDirFmt:
+    # Initialize a directory format object to store the Minimap2 index
```
Comment (VinzentRisch) [inline q2_viromics/checkv_fetch_db.py:35]:
```suggestion
    # Initialize a directory format object to store the CheckV database
```

File: q2_viromics/checkv_fetch_db.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    database = CheckVDbDirFmt()
+
+    # Construct the command to build the Minimap2 index file
```
Comment (VinzentRisch) [inline q2_viromics/checkv_fetch_db.py:38]:
```suggestion
    # Construct the command to download the CheckV database
```

File: q2_viromics/checkv_fetch_db.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        run_command(cmd)
+    except subprocess.CalledProcessError as e:
+        raise Exception(
+            "An error was encountered while running CheckV download_database, "
+            f"(return code {e.returncode}), please inspect "
+            "stdout and stderr to learn more."
+        )
+
+
+# Fetch the Virsorter2 database
```
Comment (VinzentRisch) [inline q2_viromics/checkv_fetch_db.py]:
```suggestion
# Fetch the CheckV database
```

File: q2_viromics/types/_type.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2024, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+from qiime2.plugin import SemanticType
+
+CheckVDb = SemanticType("CheckVDb")
```
Comment (VinzentRisch) [inline q2_viromics/types/_type.py]:
```suggestion
CheckVDB = SemanticType("CheckVDB")
```

File: q2_viromics/tests/data/type/db/checkVdb/hmm_db/checkv_hmms/1.hmm
Comment (VinzentRisch) [inline q2_viromics/tests/data/type/db/checkVdb/hmm_db/checkv_hmms/1.hmm:1]: Your plugin is 76.2 Mb big and 76.1 are just database files.
See how you can reduce this. Most of it is unnecessary. Like having 3 hmm files, one would be enough. Also maybe find a way to make the files smaller. They dont have to be the original files for testing purposes.

**Review by VinzentRisch — CHANGES_REQUESTED** [review]
Hi see some changes and questions below.

File: q2_viromics/tests/data/type/db/checkVdb/hmm_db/checkv_hmms/1.hmm
Comment (ChristosMatzoros) [reply q2_viromics/tests/data/type/db/checkVdb/hmm_db/checkv_hmms/1.hmm:1]: I will reduce these files.

**Comment by ChristosMatzoros** [comment]
Hello @VinzentRisch, I made the requested changes. You can have a look.

File: q2_viromics/tests/test_format.py [hunk 11]
```diff
@@ -56,6 +59,63 @@ def test_HMMFormat_neg1(self):
...
+
+    def setUp(self):
+        self.format = CheckVDBDirFmt("/dummy/path", mode="r")
+
+    @patch.object(CheckVDBDirFmt, "hmm_files_path_maker")
+    def test_hmm_files_path_maker(self, mock_hmm_files):
+        mock_hmm_files.return_value = "/dummy/path/hmm/dir/name.hmm"
+        result = self.format.hmm_files_path_maker("outer", "dir", "name")
+        mock_hmm_files.assert_called_once_with("outer", "dir", "name")
+        self.assertEqual(result, "/dummy/path/hmm/dir/name.hmm")
```
Comment (VinzentRisch) [inline q2_viromics/tests/test_format.py]: You have to explicitly state the parameters
```suggestion
    @patch.object(CheckVDBDirFmt, "hmm_files_path_maker")
    def test_hmm_files_path_maker(self, mock_hmm_files):
        mock_hmm_files.return_value = "/dummy/path/hmm/dir/name.hmm"
        result = self.format.hmm_files_path_maker(outer="outer", dir="dir", name="name")
        mock_hmm_files.assert_called_once_with("outer", "dir", "name")
        self.assertEqual(result, "/dummy/path/hmm/dir/name.hmm")
```

**Review by VinzentRisch — CHANGES_REQUESTED** [review]
Hi looks good! just fix the path maker tests.

**Comment by ChristosMatzoros** [comment]
@VinzentRisch I did the changes in the unittests.

**Review by VinzentRisch — APPROVED** [review]
Looks good! feel free to merge

**Pull request merged**